cmd/keymasterd/unseal.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/armor"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+)
+
+func (state *RuntimeState) secretInjectorHandler(w http.ResponseWriter,
+	r *http.Request) {
+	// checks this is only allowed when using TLS client certs.. all other authn
+	// mechanisms are considered invalid... for now no authz mechanisms are in
+	// place i.e. Any user with a valid cert can use this handler
+	if r.TLS == nil {
+		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		logger.Printf("We require TLS\n")
+		return
+	}
+	if len(r.TLS.VerifiedChains) < 1 {
+		state.writeFailureResponse(w, r, http.StatusForbidden, "")
+		logger.Printf("Forbidden\n")
+		return
+	}
+	clientName := r.TLS.VerifiedChains[0][0].Subject.CommonName
+	logger.Printf("Got connection from %s", clientName)
+	r.ParseForm()
+	sshCAPassword, ok := r.Form["ssh_ca_password"]
+	if !ok {
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"Invalid Post, missing data")
+		logger.Printf("missing ssh_ca_password")
+		return
+	}
+	if err := state.unsealCA([]byte(sshCAPassword[0]), clientName); err != nil {
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"Invalid Post, "+err.Error())
+		logger.Println(err)
+		return
+	}
+	w.WriteHeader(200)
+	fmt.Fprintf(w, "OK\n")
+	//fmt.Fprintf(w, "%+v\n", r.TLS)
+}
+
+func (state *RuntimeState) tryAutoUnseal() error {
+	baseConfig := state.Config.Base
+	if baseConfig.UnsealPasswordSecretArn == "" {
+		return nil
+	}
+	secretArn, err := arn.Parse(baseConfig.UnsealPasswordSecretArn)
+	if err != nil {
+		return err
+	}
+	if baseConfig.UnsealPasswordSecretKey == "" {
+		baseConfig.UnsealPasswordSecretKey = "UnsealPassword"
+	}
+	creds := credentials.NewCredentials(&ec2rolecreds.EC2RoleProvider{
+		Client:       ec2metadata.New(session.New(&aws.Config{})),
```

Comment:
I do this differently which I think is clearer:
func getAWSRegion() (string, error) {
        svc := ec2metadata.New(session.New())
        if !svc.Available() {
                return "", fmt.Errorf("not running on AWS or metadata not available")
        }
        return svc.Region()
}

and then:
    region, err := getAWSRegion()
        if err != nil {
                return nil, err
        }
        sess, err := session.NewSession(&aws.Config{
                Region: aws.String(region),
        })
        if err != nil {
                log.Printf("failed Sessionm err=%s", err)
                return nil, err
        }
        //Create a Secrets Manager client
        svc := secretsmanager.New(sess)
        input := &secretsmanager.GetSecretValueInput{
                SecretId:     aws.String(s.config.Unsealing.AWSSecretName),
        }

Reply:
Done.

Reply:
As discussed offline, I've added a TODO.

Reply:
I've simplified this as suggested and split out the core code that deals with AWS Secrets Manager so that it can be re-used for other secrets (like the database password).